File: top_interview_questions/javascript/94_binary_tree_inorder_traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var inorderTraversal = function(root) {
    let ans = [];
    inorder(ans, root);
    return ans;
};

var inorder = function(ans, root) {
        if (!root) return;
        inorder(ans, root.left);
        ans.push(root.val);
        inorder(ans, root.right);
};

var iteratingInorderTraversal = function(root) {
    let ans = [];
    let stack = [];

    while (root || stack.length) {
        while(root) {
            stack.push(root);
            root = root.left;
        }
        root = stack.pop();
        ans.push(root.val);
        root = root.right;
    }

    return ans;
};

File: top_interview_questions/python/94_binary_tree_inorder_traversal.py
#!/usr/bin/env python
# encoding: utf-8
'''
@author: Fizz Cao 
@file: 94_binary_tree_inorder_traversal.py
@time: 20/09/2021 10:21 pm
'''


# Definition for a binary tree node.
from typing import Optional, List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def inorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans = []
        self.inorder(ans, root)
        return ans

    def inorder(self, ans, root):
        if not root:
            return
        self.inorder(ans, root.left)
        ans.append(root.val)
        self.inorder(ans, root.right)

    # iterating method
    def iteratingInorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        ans, stack = [], []

        while (root or stack):
            while root:
                stack.append(root)
                root = root.left
            root = stack.pop()
            ans.append(root.val)
            root = root.right

        return ans